### ISSUE TITLE
VCS dir should only be deleted if it was copied before

### ISSUE THREAD

**Issue by KLBonn** [jenkinsci/last-changes-plugin#46]
I get an error (see stacktrace below), possibly because of a situation where:

- pipeline runs on a single Jenkins (no master-slave setup)
- "vcsTargetDir" possibly is equal to "remoteGitDir", so "vcsDirFound" did not get copied at line 153 in the first place
- the .git folder is still needed for additional custom pipeline activities afterwards

I presume most people do not do anything with their .git folder in their pipeline besides using lastChanges, so they do not notice its deletion (or run into this issue).

I think, the [delete call](https://github.com/jenkinsci/last-changes-plugin/blob/c6a2c6dafa0fcd5bac9b34f39f584db637bb60b6/src/main/java/com/github/jenkins/lastchanges/LastChangesPublisher.java#L292) should happen only if the [copy "branch"](https://github.com/jenkinsci/last-changes-plugin/blob/c6a2c6dafa0fcd5bac9b34f39f584db637bb60b6/src/main/java/com/github/jenkins/lastchanges/LastChangesPublisher.java#L153) (or line 177) was hit before.

Stacktrace:

> java.nio.file.FileSystemException: C:\Program Files (x86)\Jenkins\jobs\{{my_project}}\branches\master\builds\23\.git\objects\pack\pack-bff7ccf14f020218be91e14508750dea6ab1a157.pack: The process cannot access the file because it is being used by another process.
> 
> 	at sun.nio.fs.WindowsException.translateToIOException(Unknown Source)
> 	at sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
> 	at sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
> 	at sun.nio.fs.WindowsFileSystemProvider.implDelete(Unknown Source)
> 	at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(Unknown Source)
> 	at java.nio.file.Files.deleteIfExists(Unknown Source)
> 	at hudson.Util.tryOnceDeleteFile(Util.java:316)
> 	at hudson.Util.deleteFile(Util.java:272)
> Also:   hudson.AbortException: script returned exit code -1
> 		at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$Execution.check(DurableTaskStep.java:335)
> 		at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$Execution.run(DurableTaskStep.java:288)
> 		at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
> 		at java.util.concurrent.FutureTask.run(Unknown Source)
> 		at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(Unknown Source)
> 		at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
> Caused: java.io.IOException: Unable to delete 'C:\Program Files (x86)\Jenkins\jobs\{{my_project}}\branches\master\builds\23\.git\objects\pack\pack-bff7ccf14f020218be91e14508750dea6ab1a157.pack'. Tried 3 times (of a maximum of 3) waiting 0,1 sec between attempts.
> 	at hudson.Util.deleteFile(Util.java:277)
> 	at hudson.FilePath.deleteRecursive(FilePath.java:1322)
> 	at hudson.FilePath.deleteContentsRecursive(FilePath.java:1331)
> 	at hudson.FilePath.deleteRecursive(FilePath.java:1313)
> 	at hudson.FilePath.deleteContentsRecursive(FilePath.java:1331)
> 	at hudson.FilePath.deleteRecursive(FilePath.java:1313)
> 	at hudson.FilePath.deleteContentsRecursive(FilePath.java:1331)
> 	at hudson.FilePath.deleteRecursive(FilePath.java:1313)
> 	at hudson.FilePath.access$1600(FilePath.java:209)
> 	at hudson.FilePath$DeleteRecursive.invoke(FilePath.java:1291)
> 	at hudson.FilePath$DeleteRecursive.invoke(FilePath.java:1287)
> 	at hudson.FilePath.act(FilePath.java:1095)
> 	at hudson.FilePath.act(FilePath.java:1078)
> 	at hudson.FilePath.deleteRecursive(FilePath.java:1285)
> 	at com.github.jenkins.lastchanges.LastChangesPublisher.perform(LastChangesPublisher.java:292)
> 	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:80)
> 	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
> 	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1$1.call(SynchronousNonBlockingStepExecution.java:49)
> 	at hudson.security.ACL.impersonate(ACL.java:290)
> 	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1.run(SynchronousNonBlockingStepExecution.java:46)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
> 	at java.util.concurrent.FutureTask.run(Unknown Source)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
> 	at java.lang.Thread.run(Unknown Source)
> Finished: FAILURE
>

**Comment by rmpestano** [comment]
Hi @KLBonn, I got it.

Do you have any sample I can reproduce this situation? Any plugin I should run after last-changes? 

Thank you.

**Comment by KLBonn** [comment]
Hi @rmpestano ,

sorry, I failed to extract our setup to a standalone pipeline :-(

We have some shared pipeline library, which calls lastChanges, and then uses org.eclipse.jgit to do further diff calculations (like calculating  amount of change / percentage of changed code lines etc.).
I tried, but could not realize that logic (with dependency on jgit) in a simple scripted pipeline.

Based on some pipeline exampe, the general idea looks like:

```
@Grab(group='org.eclipse.jgit', module='org.eclipse.jgit', version='4.11.0.201803080745-r')
import org.eclipse.jgit.storage.file.FileRepositoryBuilder

node {
   def mvnHome
   stage('Preparation') { // for display purposes
      // Get some code from a GitHub repository
      git 'https://github.com/jglick/simple-maven-project-with-tests.git'
      // Get the Maven tool.
      // ** NOTE: This 'M3' Maven tool must be configured
      // **       in the global configuration.           
      mvnHome = tool 'Maven 3.x'
   }
   stage('Build') {
       script {
           lastChanges format:'SIDE',matching: 'WORD'
          def builder = new org.eclipse.jgit.storage.file.FileRepositoryBuilderFileRepositoryBuilder()
          def repository = builder.setGitDir(new File("${pwd()}/.git")).build()
           builder = null
            assert repository
            def git = new org.eclipse.jgit.api.Git(repository)
            def commits = git.log().all().call()
            for (commit in commits) {
                echo "Commit: ${commit.shortMessage}"
            }
            // or do some other foo with Git SCM info
        }
   }
   stage('Results') {
      junit '**/target/surefire-reports/TEST-*.xml'
   }
   stage('Cleanup') {
       deleteDir()
   }
}
```

**Comment by rmpestano** [comment]
Maybe we could add a parameter to skip VCS dir cleanup, WDYT?

**Comment by KLBonn** [comment]
Step argument or flag?

I thought, a simple solution could be some local variable for memorizing if vcsTargetDir is result of a copy (modification only in LastChangesPublisher.java):
Like, `private boolean isCopiedVcsDir = false;` (defined below isSvn at line 110).
Then set it true inside the ìf` blocks at lines 153 and 177, and include it in the ìf` condition at line 291.

**Comment by IsharaRuchiranga** [comment]
I'm facing the same issue in pipeline. @KLBonn do you have any workaround for this issue?

**Comment by KLBonn** [comment]
Sorry for not answering earlier.
My job focus moved on to other topics, so I could not afford spending more time with this issue.

It seems only Windows is plagued by this issue (on Linux the filesystem will just defer deletion until it can be done). And since Windows is no primary supported OS for Jenkins at our site anyway, we just skipped last-changes plugin on those instances.

**Comment by rmpestano** [comment]
The fix is available in **v2.7.0**.
